c2_utils/src/mfx_gralloc4.cpp

```diff
@@ -0,0 +1,224 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "mfx_gralloc4.h"
+#include "cros_gralloc/cros_gralloc_helpers.h"
+#include "cros_gralloc/i915_private_android.h"
+#include "mfx_c2_utils.h"
+#include "mfx_debug.h"
+#include "mfx_c2_debug.h"
+
+#undef MFX_DEBUG_MODULE_NAME
+#define MFX_DEBUG_MODULE_NAME "mfx_gralloc4"
+
+using aidl::android::hardware::graphics::common::PlaneLayout;
+
+c2_status_t MfxGralloc4Module::Init()
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    sp<IMapper4> mapper = IMapper4::getService();
+    if (nullptr == mapper)
+    {
+        return C2_CORRUPTED;
+    }
+
+    m_mapper = mapper;
+    
+    return C2_OK;
+}
+
+MfxGralloc4Module::~MfxGralloc4Module()
+{
+
+}
+
+Error4 MfxGralloc4Module::get(const native_handle_t* handle, const IMapper4::MetadataType& metadataType,
+                    hidl_vec<uint8_t>& outVec)
+{
+    Error4 err;
+    if (nullptr == m_mapper)
+        return Error4::NO_RESOURCES;
+    m_mapper->get(const_cast<native_handle_t*>(handle), metadataType,
+                [&](const auto& tmpError, const hidl_vec<uint8_t>& tmpVec)
+                {
+                    err = tmpError;
+                    outVec = tmpVec;
+                });
+    return err;
+}
+
+c2_status_t MfxGralloc4Module::GetBufferDetails(const buffer_handle_t handle, BufferDetails *details)
+{
+    MFX_DEBUG_TRACE_FUNC;
+    c2_status_t res = C2_OK;
+
+    do
+    {
+        details->handle = handle;
+
+        auto cros_hnd = cros_gralloc_convert_handle(handle);
+        details->prime = cros_hnd->fds[0];
+        MFX_DEBUG_TRACE_I32(details->prime);
+
+        hidl_vec<uint8_t> vec;
+
+        if (IsFailed(get(handle, gralloc4::MetadataType_Width, vec)))
+        {
+            res = C2_CORRUPTED;
+            break;
+        }
+
+        uint64_t width = 0;
+        gralloc4::decodeWidth(vec, &width);
+        details->width = details->allocWidth = width;
+        MFX_DEBUG_TRACE_I32(details->width);
+
+        if (IsFailed(get(handle, gralloc4::MetadataType_Height, vec)))
+        {
+            res = C2_CORRUPTED;
+            break;
+        }
+
+        uint64_t height = 0;
+        gralloc4::decodeHeight(vec, &height);
+        details->height = details->allocHeight = height;
+        MFX_DEBUG_TRACE_I32(details->width);
+
+        uint32_t drm_format = 0;
+        if (IsFailed(get(handle, gralloc4::MetadataType_PixelFormatFourCC, vec)))
+        {
+            res = C2_CORRUPTED;
+            break;
+        }
+        gralloc4::decodePixelFormatFourCC(vec, &drm_format);
+        auto format = i915_private_invert_format(drm_format);
+        details->format = format;
```

Comment:
Could this format be got from MetadataType_PixelFormatRequested? Instead of directly call minigbm private functions "i915_private_invert_format()"

Reply:
> Could this format be got from MetadataType_PixelFormatRequested? Instead of directly call minigbm private functions "i915_private_invert_format()"

hi chenyan, thx for your suggestion, I tried this type, it works well.